### ISSUE TITLE
Integrate interp-engine as the first hub-available adapter module (Phase 0 complete: GO)

### ISSUE THREAD

**Issue by speediedan** [speediedan/interpretune#372]
Make [interp-engine](https://github.com/decoderesearch/interp-engine) the first hub-available adapter module: the adapter analogue of what the jlens collection is for non-bundled ops. It is a genuinely third-party package, from the Neuronpedia collaborators, which post-refactor neuronpedia already depends on directly (`interp-engine==1.3.3` in `apps/inference`). That makes it the strongest available evidence that hub component kinds beyond ops are a real ecosystem surface, and it fires #125's own deferral condition ("until a concrete third-party adapter need exists").

## Phase 0 is done, and the verdict is GO

Validation spike run 2026-08-26 in a scratch env beside the pinned stack, CPU only, no pinned dependency touched and no resolver movement.

- `EagerModel(id, hf_model=existing)` wraps an already-loaded HF model in place, with no reload and no second copy. Interpretune's `auto_model_init` / `post_auto_model_init` lifecycle composes with it directly, the same shape as the TL and nnsight adapters.
- The sync free functions work from sync code, so `*_step` and analysis-op call sites bind to them without an event loop.
- Capture is TransformerLens-exact where the names denote the same tensor: `resid_pre`, `resid_mid`, `resid_post` at cosine 1.000000 (rel_l2 ~1e-7) on both gpt2 and gemma-3-1b-it.
- The sandwich-norm trap behaves exactly as upstream documents, and is now quantified on the architecture our own corpora use. gemma-3-1b-it layer 5: passing the model resolves `hook_mlp_out` to `mlp_out_post` and matches TL at cosine 1.000005, against 0.109 for the string-only mapping; `hook_attn_out` gives 1.000004 against 0.221. Upstream publishes 0.2 to 0.4 from gemma-2, so gemma-3 is worse than the range they quote.

One mapping mismatch was found and chased to a conclusion: interp-engine maps TL's `hook_mlp_in` to its post-norm `mlp_in`, and those are different tensors (cosine 0.088 on gemma-3-1b-it). Measuring which one the artifacts actually want settled it against TransformerLens rather than against interp-engine: Google's shipped transcoder config declares `pre_feedforward_layernorm.output`, and reconstruction gives FVU 0.11 at L0 17 from that tensor against FVU 20585 at L0 819 from `hook_mlp_in`, with a declared L0 of 15. So interp-engine reads the tensor the artifact wants, and the defect is in SAELens' declared hook name. That is being reported upstream separately and does not block this work.

The practical consequence for the adapter is a design rule rather than a blocker: **the seam module normalizes block-level sublayer-input hooks itself rather than trusting a name round-trip**, and hook translation is always model-aware.

## Scope

**Phase 1: the #125 rails** (tracked on #125, unchanged scope). `adapters:` manifest section, trust gate to dynamic module to `CompositionRegistry.register`, `requires:` enforcement, card tags. Tested first against a trivial in-repo fixture adapter, so rails and passenger fail independently.

**Phase 2: the adapter itself.** Eager tier only.

- Surface A, the hub-delivered artifact: `Adapter.interp_engine` registered dynamically, `InterpEngineConfig`, a mixin wrapping `self.model` in `EagerModel` post-init, and one internal seam module owning every `interp_engine` import and symbol. Composition targets: `(core, interp_engine)` and `(lightning, interp_engine)`.
- ~~Surface B, a named `AnalysisBackend` entry~~ **moved to Phase 3.** See the re-sequencing note below.
- Capability-gated tests that skip rather than fail where a capability is absent, `RunIf`-gated for CUDA paths.
- A registered example composition and an interp-engine version of the concept-steering demo, converted to CLI form if the notebook path proves unworkable. The sync free functions refuse inside a running event loop and Jupyter has one, so the notebook leads with the plan's hardest known problem; the CLI fallback exists for that.

**Phase 3, deferred:** Surface B (the `AnalysisBackend` entry, hub-delivered; see the condition below), vLLM tier-2 backend value, `capture_attention`, steering-spec alignment with the jlens and concept ops, `decode_residuals` against the unembed and final-norm-scale seam (#225/#333), and benchmark-lane evaluation of the static-tap decode speedup.

## Decisions taken

- Eager tier first. vLLM tier 2 only, later; tier 1 never, since we do not own a vLLM engine and tier 2 owns the correctness details tier 1 leaves to callers.
- Pin `interp-engine ~=1.5.1; python_version >= '3.11'` and keep every engine call behind one seam
  module. **The marker is required, not cosmetic** (interp-engine supports >=3.11 while this project
  supports 3.10), and `pyproject.toml` says so at the pin site; an earlier revision of this line
  recorded `~= 1.3` with no marker, which is the wrong version *and* silently drops a constraint the
  tree documents as load-bearing. The upstream 1.x line changes signatures without deprecation aliases (1.1 and 1.3 both did), deliberately, preferring load-time errors over silently wrong engines. PyPI moved 1.3.4 to 1.3.6 within two days of the assessment, so this guard is load-bearing rather than precautionary.
- Hook translation is always model-aware; string-only translation is banned in review.
- Capability gating by query and refusal (`CAPABILITIES`, `CapabilityUnsupported`), never by backend name.
- The adapter component repo lives under `speediedan/` initially. Offering to move it to `decoderesearch/` is a later conversation, once its shape is settled.
- This work does **not** gate the neuronpedia #217 push. That rebase is perishable and its delay cost is measurable, where the goodwill from an adapter is diffuse. If a working slice exists first, it gets promoted and referenced there; that is upside, not a dependency.

## Surface B moves to Phase 3; Phase 2 closes on Surface A alone (MAINTAINER decision)

**Decided by the maintainer**, ratifying a descope proposed from the interp-engine lane. Attribution
stated because this repository's shared account means a comment's byline says nothing about who decided
its content, and the difference matters here: the measurement below is re-takeable by anyone, the
re-sequencing is not.

**The condition that would make Surface B real**, stated instead of only the phase it moved to, because
a phase label ages into "why is this here" while a condition can be evaluated:

> Surface B becomes buildable when interpretune has analysis ops needing the layer **above** execution
> AND interp-engine can serve them. Concretely, either interp-engine grows an attribution or
> feature-intervention capability, **or the `AnalysisBackendCapability` vocabulary grows a member
> interp-engine implements.**

**No concrete candidate is named, deliberately.** An earlier revision of this section pointed at #450's
capturable-subset axis as the likely route. That was wrong and is removed rather than qualified, because
a wrong pointer reads as a lead: that axis grows the **model-backend** contract, a support record beside
`InterventionSupport`, since `fwd_w_cache` is a base method every model backend has. Surface B's
emptiness is about **`AnalysisBackendCapability`**, a different enum on a different surface. When a
candidate exists it can be named here.

**Why it is not buildable today**, in one line with the measurement recorded in a comment on this issue:
every bundled op reaches execution through `model_backend.*`, which the adapter already implements;
`AnalysisBackendCapability` has two members and interp-engine implements neither; and the only consumer
of the analysis-backend seam is `circuit_tracer/ct_ops.py`. A backend written now would declare an empty
capability set and stub fourteen methods nothing calls, so **the assertion that would make it reviewable
is one whose negative case cannot run.**

**Placement, when it is built: hub-delivered, not in-tree.** This supersedes an earlier in-tree position
recorded here, which was argued from `circuit_tracer`'s precedent and from the flags being low-risk;
neither premise held. The deciding argument is enforceability: the adapter's `_Seam` is documented as
*"The one place this adapter touches `interp_engine`."* An in-tree backend would import engine symbols
from a second place in another repository, where no test in either could enforce that invariant. It
would not merely weaken; it would become unfalsifiable. That invariant is now enforced by a test in the
adapter repo, so the argument rests on something checkable rather than on an assumption.

### Where Surface B is tested, since hub-delivered must not become untested

Recorded as a decision rather than left implicit, because interpretune cannot exercise a hub-delivered
backend without pulling the component.

- **The absent case is already covered in-tree, and generically.**
  `tests/core/test_backend_name_resolution.py` asserts
  `KeyError, match="No analysis backend registered as 'not_a_backend'"`. It names no component, so it
  holds whether or not interp-engine is installed, and it is the negative control that the refusal path
  still refuses by name.
- **Protocol satisfaction is asserted adapter-side**, since interpretune cannot import the component to
  check it. `AnalysisBackend` is `@runtime_checkable`, so the assertion is available there.
- **A caveat for whoever writes it.** A bare `isinstance(backend, AnalysisBackend)` is fragile under a
  suite that loads the capabilities module twice, which is a known failure mode here: the adapter's own
  `supports` already compares capabilities **by value** for exactly that reason. Prefer a structural
  assertion or a by-value comparison over relying on identity.

Cross-reference: #125 (the rails this rides), #231 (the Scalable Dashboards wave these collaborators run).

**Comment by speediedan** [comment]
## Two findings against the pinned surface, and a recommendation on the open question

Measured against `interp-engine 1.5.1` as the pins produce it (`INSTALLER: uv`, no `direct_url.json`),
not a checkout.

### 1. `CAPABILITIES` is a table of what vLLM *cannot* do, so an eager-only Phase 2 derives nothing from it

Phase 2's Surface B is scoped as "a named `AnalysisBackend` entry with capability flags derived from
`interp_engine.CAPABILITIES`", and Phase 2 is eager tier only. Those two do not compose, and the reason
is in the upstream docstring rather than in inference:

> Every capability **one backend has and the other does not**.

Every one of the six rows is a vLLM limitation, and each `why` names the mechanism: the unembed runs in
a worker through vLLM's own `compute_logits`; vLLM owns its weights in a worker subprocess; this arm
forwards one unpadded sequence. The enforcement helper is `require_eager(model, what, capability=...)`,
which *refuses unless the model is eager*.

So on the eager tier every row in that table is **available**. Deriving flags from it would produce a
capability table asserting nothing, and a capability-gated test suite that skips nothing and therefore
cannot fail. That is the shape this repo already treats as a defect: a guard whose negative case never
runs reports success forever.

**What Surface B's flags should come from instead**: interpretune's own `AnalysisBackendCapability`
semantics, set from what the eager arm actually does, measured. `CAPABILITIES` becomes the right source
exactly when the vLLM tier lands in Phase 3, which is also when the flags stop being vacuous. Worth
re-scoping the bullet now rather than discovering it while writing the tests.

### 2. The recorded pin has drifted from the tree

Decisions taken says **`interp-engine ~= 1.3`**; `pyproject.toml` carries
**`interp-engine ~=1.5.1; python_version >= '3.11'`**. The reasoning behind the decision is unaffected
and if anything strengthened, since the 1.x line kept moving exactly as predicted. But the recorded
number is the one a reader will quote, and the marker (which is load-bearing, not cosmetic) is not
recorded here at all.

### 3. The open question: both placements are already supported, so the choice is about skew

> Does Surface B's backend entry land in-tree, like existing backends, or inside the hub adapter repo?

The machinery already serves both, which the resolver says in its own docstring:

> Bundled backends register themselves at import time... Hub-delivered backends register from their
> component entrypoint and are already present by the time anything resolves them, so they never reach
> the table.

So this is not a capability question. Two things that do bear on it:

- **The absent-implementation case is already correct for the in-tree path.** `resolve_analysis_backend`
  lazily imports via `_BUNDLED_BACKEND_MODULES`, swallows the `ImportError`, and then refuses by name:
  `No analysis backend registered as 'x' (known: [...]). Hydrating this artifact requires the backend's
  package/extra to be installed.` An in-tree entry for an engine nobody has installed therefore fails
  loudly and actionably rather than silently, which is the property that usually argues against
  in-tree entries for optional dependencies.
- **The argument against in-tree is version skew, not mechanism.** In-tree capability flags describe an
  engine version, while the hub component that supplies the adapter carries its own pin. Split them and
  the flags can describe an engine the loaded component does not match, with nothing to detect it.

**Recommendation: in-tree for the first slice, as the issue already leans**, following `circuit_tracer`'s
exact pattern (an entry in `_BUNDLED_BACKEND_MODULES`, registration at module import, no engine import at
interpretune import time so the import-time guards stay green). The skew risk is real but small while the
flags are eager-only and, per finding 1, largely vacuous. **Revisit when the vLLM tier lands**, because
that is simultaneously when the flags start asserting something and when skew starts to cost.

**Comment by speediedan** [comment]
## Measurement: Surface B is not currently buildable in a form that could be tested

Recorded as measurement rather than as a scope change. Every number here is re-takeable from the
commands beside it; the question of whether Phase 2 re-sequences around it is not settled here.

### Ops reach capture and steering through `ModelBackend`, not `AnalysisBackend`

Surface B is justified in this issue as *"so ops reach capture and steering through the existing backend
seam rather than importing engine code into op definitions."* Every bundled op reaches execution through
`model_backend.*`:

```
core_ops.py:97          model_backend.fwd_w_cache(...)
concept_ops.py:1038     model_backend.fwd_w_intervention(...)
sae_ops.py:83,147       fwd_w_cache_and_latent_models / fwd_w_hooks_batched
attribution_ops.py:74   fwd_w_grads_and_latent_models
attribution_ops.py:134  wrap_activation_cache
```

`AnalysisBackend` is described in its own docstring as "analysis-adapter functionality layered **above**
model execution backends". The adapter already ships `InterpEngineModelBackend` implementing
`ModelBackendCore` plus `SupportsIntervention`.

### The capability vocabulary has two members and this component implements neither

`AnalysisBackendCapability` is exactly `ATTRIBUTION_GRAPH` and `FEATURE_INTERVENTION`. An interp-engine
backend would declare `frozenset()`.

### Consumer side

The only file under `src/interpretune/analysis/ops/bundled/` referencing an analysis backend is
`circuit_tracer/ct_ops.py`.

### The protocol is 16 methods, and the two that are reachable are overrides rather than requirements

The surface is `capabilities`, `supports`, `get_tokenizer`, `get_embedding_weight`,
`token_strings_to_ids`, `resolve_prompt`, `build_concept_attribution_targets`,
`resolve_feature_intervention_settings`, `build_feature_interventions`,
`feature_intervention_call_kwargs`, `decompose_graph`, `hydrate_graph_from_batch`, `build_pruned_graph`,
`select_feature_rows`, `compute_node_influence_scores`, `compute_signed_node_influence_scores`.

**Correcting a narrower claim made while assessing this**: `get_tokenizer` and `get_embedding_weight`
*are* called — not from op files, but from op infrastructure, at `optools.py` in `resolve_tokenizer` and
`resolve_embedding_weight`. "Called by no op file" was true and was answering a narrower question than
"is this method reachable".

Reading them shows they do not change the conclusion. Both call the analysis backend inside a `try`,
catch `AttributeError` / `ValueError`, and fall through to an attribute-path chain (`model.tokenizer`,
`datamodule.tokenizer`; `replacement_model.unembed_weight`, `model.W_E`, and others). With no
interp-engine analysis backend registered, `get_analysis_backend` returns `None` and the fallback path
already handles it. **They are overrides, not requirements.**

`token_strings_to_ids` has no call sites: the `optools.py` matches are a standalone free function of the
same name and its `__all__` entry, not calls to the protocol method.

### What that adds up to

An interp-engine `AnalysisBackend` written today would declare an empty capability set, stub fourteen
methods nothing calls, and override two that already resolve without it. **A protocol-satisfaction
assertion on an object no caller invokes is a guard whose negative case cannot run**, so the thing that
would make it reviewable is also the thing that cannot be written yet.

Re-sequencing is a separate decision and is not taken here.

**Comment by speediedan** [comment]
The abstraction question raised here, the two capability enums governed by different rules and sharing the word "intervention" across two objects, now has its own issue with the proposal: #537. It records the measurement that bears on this issue's condition: the circuit-tracer analysis backend executes on its own replacement model (a `CT_BACKEND_REGISTRY`-keyed factory product) and never through `module.model_backend`, so a composition of interp-engine with circuit-tracer keeps the analysis capabilities circuit-tracer's, and interp-engine underneath it would be a registry entry rather than an `AnalysisBackend` entry. That is consistent with the descope above and makes the open condition concrete.

**Comment by speediedan** [comment]
## Scoping status 2026-09-23 (verified against origin/main `d78702b8`)

**Closed since last review:** #554 is closed (PR #606 merged as `e679546c`); the docs gallery-row fix for this lane's notebook merged as `d78702b8`. Both landed clean.

**Claim-base staleness immaterial:** the hold's base (`cd24ff83`) is four merges behind main (#421, #338, #606, #616; 17 files), and none touch interp-engine/hub-adapter/component paths — the analysis-dir touches are jlens ops and conformance/runner files. No re-claim needed.

**Remaining open set:** #125 (rails), #396 (requires floor), #450 (conformance suite), #577 (utils promotion) — all owned outside this issue's scope as previously scoped. #427 (J-space section) is being worked now; the N2c-7 upstream step stays gated on its evidence plus the dashboard set.

**#401 closure-verification (needs routing):** checked against `origin/main`. Criterion 1 holds — no adapter-named module remains under `config/`, and all four per-adapter packages carry `config.py`. Criterion 2 needs a design ruling, not more tree reads: `register_module_cfg_class` exists as the hub route, but bundled adapters still resolve via import-path discovery first (registering changes nothing for them, per the docstring), so whether that satisfies "the same mechanism" is a judgment call. Criterion 3 (test gate) is unverified — execution is deferred while the Azure pipeline job holds this host. This needs a route to the design owner; I have no offer path there, so treating this comment as the routing request.

**Net:** no executable remainder in this issue tonight. Holding for (a) the N2c-7 evidence gate and (b) review of the scoping summary above.
